### PR TITLE
feat: allow changing password validation API DNS name

### PR DESCRIPTION
**Pull request by radekg**
## Related issue

This PR relates to https://github.com/ory/kratos/issues/969.

## Proposed changes

Allow changing the DNS name used for validating the password exposure.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

## Further comments

By design, I do expect the replacement API to always use TLS. Currently, there is no option to set a custom TLS certificate for the client used in the validator. It would be awesome to allow custom certificate for an updated host but I'm not exactly sure what would be the best way to do so. I think, the best would be to fetch the config in https://github.com/radekg/kratos/blob/12a62ac03253f56a22a5938b503fee3ca8d4132b/selfservice/strategy/password/validator.go#L69 and create the TLSContext there but how do I pass the context to the roundtripper created via the `httpx.NewResilientClientLatencyToleranceMedium`?

Without an option to set a custom certificate, the usual mess of having to add a custom CA chain to the OS is needed. Being able to set a custom cert, key and optionally a CA chain would be preferred.

Additionally, please point me to what changes are required in the documentation bits.

### CODE REVIEW

File: driver/config/config.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ const (
 	ViperKeyHasherArgon2ConfigParallelism                           = "hashers.argon2.parallelism"
 	ViperKeyHasherArgon2ConfigSaltLength                            = "hashers.argon2.salt_length"
 	ViperKeyHasherArgon2ConfigKeyLength                             = "hashers.argon2.key_length"
+	ViperKeyPasswordAPIDNSName                                      = "selfservice.methods.password.config.api_dns_name"
```
Comment (aeneasr) [inline driver/config/config.go:85]: Can we rename this to:

```suggestion
	ViperKeyPasswordAPIDNSName                                      = "selfservice.methods.password.config. haveibeenpwned_url"
```

**Review by aeneasr — COMMENTED** [review]
Awesome, thank you for your contribution! This looks pretty good but I have some ideas how to improve it further :)

**Comment by aeneasr** [comment]
Are you still up for the changes? :) If you need any help, let us know!

**Comment by radekg** [comment]
Hey @aeneasr. Sorry for the delay. I'm not sure if I'm fully okay with that name change, although I do agree that the original name wasn't the best. Semantically, `haveibeenpwned_url` would indicate that the complete URL can be changed but my PR allows changing the domain name (well, FQDN) only. Maybe `haveibeenpwned_api_fqdn` would be a better fit?

**Comment by aeneasr** [comment]
Yes, you're correct! `haveibeenpwned_host` is also an option!

**Comment by radekg** [comment]
Hey @aeneasr, I have updated the code.

**Review by aeneasr — COMMENTED** [review]
Great, thank you! From a code perspective, this looks good to merge! Now we need two more things:

1. Add the new config option to the [config schema](https://github.com/ory/kratos/blob/master/.schema/config.schema.json) - without that you can not actually configure this as validation will fail :)
2. Add a short guide on this - create a file `hosting-own-have-i-been-pwned-api.mdx` with title `Hosting Haveibeenpwnd API Yourself` in https://github.com/ory/kratos/tree/master/docs/docs/guides which explains how to use this feature!

Thank you for your great work!

**Comment by aeneasr** [comment]
Are you still up for the changes? :) If you need any help, let us know!

**Comment by radekg** [comment]
Will add a guide over the weekend.

**Comment by aeneasr** [comment]
Let me know when this is good for a look!

**Comment by radekg** [comment]
Hey @aeneasr, it's in progress: https://github.com/radekg/hibp.

**Comment by aeneasr** [comment]
Awesome! :) Should we keep this PR around in the meanwhile or do you want to recreate it when https://github.com/radekg/hibp is ready to be used (or is it ready to be used already?)

**Comment by radekg** [comment]
Ha, please give me one more week to wrap this up. I’m doing it just to get this pr in :) it’s only an example of how one would host their own hibp...

**Comment by aeneasr** [comment]
Ok cool :) Looking forward! :)

**Comment by radekg** [comment]
Functionality is now there. The readme is missing a reverse proxy setup, which I'll add over the next couple of days so that will cover the hard public CA requirement. I'm really sorry this took so long. Exposing this without releasing anything confidential took some working around.

**Comment by radekg** [comment]
@aeneasr I've added Traefik configuration for serving the hibp api behind the reverse proxy with ACME certs: https://github.com/radekg/hibp#setting-up-behind-reverse-proxy-with-tls.

Do you have a preference how would you like to have it available in the documentation? It's quite involved and I am almost certain you don't really want to maintain this. What do you think would be the best way to proceed here for you?

**Comment by aeneasr** [comment]
Cool stuff! I think it would probably make sense to just link to the project's documentation?

**Comment by radekg** [comment]
All right, I have added the guide. I am not sure if I have to update the sidebars manually or is that somehow auto-generated.

**Review by aeneasr — APPROVED** [review]
Awesome, thank you! 🎉 Your contribution makes Ory better :)

**Comment by aeneasr** [comment]
@radekg I get some test failures in driver/config. To resolve the formatting issues please also run `make format` in root!

**Comment by radekg** [comment]
@aeneasr, FYI, tests are passing now.

File: docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+API hosted at haveibeenpwned.com is restricted via network policies, for
+example as a result of CISO policies, you can host the HiBP API in-house
+and configure Kratos to use your own instance of the API. To do so, configure
+Kratos to use your API:
+
+```yaml title="path/to/kratos/config.yml
+selfservice:
+  methods:
+    password"
+      config"
```
Comment (zepatrik) [inline docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx]:
```suggestion
    password:
      config:
```

File: docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+:::note
+
+Ory Kratos `haveibeenpwned_host` setting does not allow configuring the
+respective CA chain at the moment but it does require a TLS connection.
+Therefore your private HiBP API must use one of the trusted CAs of the
+underlying operating system.
+
+A solution might be to host the HiBP API using a Let's Encrypt certificate
+or install the private CA certificate in the trusted CA certificates
+of the system on which Kratos is running.
```
Comment (zepatrik) [inline docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx]: I assume the note ends here?

```suggestion
of the system on which Kratos is running.

:::
```

**Review by zepatrik — COMMENTED** [review]
Some typo stuff :wink:

**Review by aeneasr — APPROVED** [review]

File: docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx (hunk 2 above)
Comment (radekg) [reply docs/docs/guides/hosting-own-have-i-been-pwned-api.mdx]: It’s been fixed in this commit: https://github.com/ory/kratos/pull/1009/commits/fbf2bc7537d9598e704e92b2b9e35cb970107aae

**Comment by aeneasr** [comment]
Awesome, great job! Thank you 😄

**Pull request merged**